### ISSUE TITLE
Parallelism

### ISSUE THREAD

**Issue by bencassie** [roadrunnerlenny/etlbox#30]
Hi there,

I have a question with regards to the level of parallelism with the DataFlow tasks.

I have 3 DBSource sql server queries, 2 of which are combined in a lookup and the result sent to a VoidDestinations whilst testing.

I notice however that the 3 DBSources seem to execute one at a time rather than in parallel. I tried putting the ExecuteAsync calls into Parallel.Invoke but no dice.

Any pointers? I see you're using TPL dataflows, am I supposed to interact with them?

Kind regards,

**Comment by roadrunnerlenny** [comment]
Hi bencassie,

Unfortunately, I wasn't able to fully reproduce your issue. But I am aware that the current implementation using Execute() / Wait() is not fully supporting an asynchronous approach.

Let me explain the current implementation first:
The `Execute()` method on a source basically will synchronously read from the source (e.g. for `DBSource` from a table) and post the records into the dataflow pipeline. This method will block until all data was read from the source and written in the flow. (Please note that depending on your flow it will need some more time until all data is transferred into the destinations). 
After all data is posted into the flow, the Execute() will return. But not all data has arrived in the destinations yet. To wait for all data to written into the destinations, you can call `Wait()` on a destination and this method will block until all records arrived at the destination. 

Initially, this looked to me as an approach which is easy to understand and write, without getting to deep into the problems of parallel programming (which also will create some more issues if not done right). In the current implementation, there is also a method called `ExecuteAsync`, but it is actually just calling the `Execute()` method. 

But as I understand, you are using two sources within a `Lookup` component. The sources are initially loaded when the dataflow starts. As the lookups internally has it's own simple dataflow containing a source and destination, it will call the Execute() method on the source and wait() for the data to be loaded. It could be that some blocking will take place here due to the synchronous setup of the Execute() / Wait() methods.  

So I was thinking of having a real "ExecuteAsync" method, but currently I am not quite happy with the naming. I am thinking of renaming the Execute to PostAll and add a new method called "StartPostAll" which starts writing data in the flow without blocking. It will also return a Task object and contain the async keyword - so not so easy to use for non experienced programmers anymore. 

I uploaded a pre-release of 1.6.2 (1.6.2-alpha) into nuget. (https://www.nuget.org/packages/ETLBox/1.6.2-alpha) It contains already an implementation for the method `StartPostAll` which can be used instead of the `Execute` method and is a fully async implementation. In this pre-release, it will be also used by the Lookup task, so it could be that this already could help you with your parallelization issue.

Other than that, using the TPL (on top of the TPL.Dataflow) is a little bit tricky and the implementations lacks of a good exception handling - whenever your dataflow throws an exception within the StartPostAll method, I wasn't able to catch that. So for testing your flow with this pre-release, please use the `Execute()` method, once the execution is working you can switch to `StartPostAll()` and having it asnychronously.

If this is not helping with your issue, I would be great if you could post a simplified version of your source code here so that I will be able to reproduce your issue.

Of course, if you have any input regarding the naming of the methods, or how you would expect the methods (synchronous and asynchrounous) to behave, this would be highly appreciated.

**Comment by bencassie** [comment]
Hi, many thanks for your reply confirming the current state and your pre-release branch.

I will get back to you when I can provide meaningful input :)

**Comment by roadrunnerlenny** [comment]
I just released version 1.6.2 (https://www.nuget.org/packages/ETLBox/1.6.2) which now has new methods to support parallelism:
All dataflow sources now support `async Task ExecuteAsync()`.
All dataflow destinations now have `asny Task Completion()`

Calling the ExecuteAsync() will return a new Task. The Task is completed when all data from the source is posted into the dataflow. 
Calling the Completion() method will also return a new Task. This task is completed when all data arrived at the destination.
You don't have to necessarily wait for the task from the source, this is only if you want to know when all data was read from the source. 

Be careful with Exception & Exception handling, ExecuteAsnyc() will start reading the data in a new Task - whenever something goes wrong here, be sure to capture this. 

Here is an example code how I used it in my test:
```
                 CustomSource source = new CustomSource(
                    () => throw new ETLBoxException("Test Exception"), () => false
               );
                 CustomDestination dest = new CustomDestination
                (row => {; });

                source.LinkTo(dest);

                Task sourceT = source.ExecuteAsync();
                Task destT = dest.Completion();
                try
                {
                    sourceT.Wait();
                    destT.Wait();
                } catch (Exception e)
                {
                    throw e.InnerException;
                }
```
